docs/source/notes/extending.rst

```diff
@@ -1,10 +1,238 @@
 Extending PyTorch
 =================
 
-In this note we'll cover ways of extending :mod:`torch.nn`,
+In this note we'll cover ways of extending :mod:`torch`, :mod:`torch.nn`,
 :mod:`torch.autograd`, and writing custom C extensions utilizing our C
 libraries.
 
+Extending :mod:`torch`
+----------------------
+
+The functions in the :mod:`torch` namespace can be overrided with custom
+user-specified implementations via the `__torch_function__` protocol introduced
+in PyTorch 1.4. Custom Python types with a `__torch_function__` method will
+bypass the standard PyTorch dispatch for :class:`Tensor` types. This works with
+"duck" types that are unrelated to :class:`Tensor` as well as user-defined
+subclasses of :class:`Tensor`.
+
+Extending :mod:`torch` with a :class:`Tensor`-like type
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. note:: This functionality is inspired by the NumPy ``__array_function__``
+          protocol.
+
+To make this concrete, let's begin with a simple example that illustrates the
+API dispatch mechanism. We'll create a custom :class:`Tensor`-like type that
+represents a 2D scalar tensor, parametrized by the order ``N`` and value
+along the diagonal entries, ``value``::
+
+  >>> class ScalarTensor(object):
+  ...  def __init__(self, N, value):
+  ...      self._N = N
+  ...      self._value = value
+  ...
+  ...  def __repr__(self):
+  ...      return "DiagonalTensor(N={}, value={})".format(self._N, self._value)
+  ...
+  ...  def tensor(self):
+  ...      return self._value * torch.eye(self._N)
+
+This first iteration of the design isn't very useful. The main functionality of
+``ScalarTensor`` is to provide a more compact string representation of a scalar
+tensor than in the base tensor class::
+
+  >>> d = ScalarTensor(5, 2)
+  >>> d
+  ScalarTensor(N=5, value=2)
+  >>> d.tensor()
+  tensor([[2., 0., 0., 0., 0.],
+          [0., 2., 0., 0., 0.],
+          [0., 0., 2., 0., 0.],
+          [0., 0., 0., 2., 0.],
+          [0., 0., 0., 0., 2.]])
+
+If we try to use this object with the :mod:`torch` API, we will run
+into issues::
+
+  >>> import torch
+  >>> torch.mean(d)
+  TypeError: mean(): argument 'input' (position 1) must be Tensor, not ScalarTensor
+
+Adding a ``__torch_function__`` implementation to ``ScalarTensor`` makes it
+possible for the above operation to succeed. Let's re-do our implementation,
+this time adding a ``__torch_function__`` implementation::
+
+  >>> HANDLED_FUNCTIONS = {}
```

Comment:
After spending time talking to @hameerabbasi in #28361, the fact that I have to make this `HANDLED_FUNCTIONS` handler struct is such a stick up my craw, haha. If only there were a way to just define these functions as methods directly... (No changes necessary here.)